src/Pact/Eval.hs

```diff
@@ -222,35 +242,156 @@ loadModule m bod1 mi g0 = do
+      do
+        d' <- forM d $ \(f :: Name) -> do
+                dm <- resolveRef f
+                case (dm,f) of
+                  (Just t,_) -> return (Right t)
+                  (Nothing,Name fn _) ->
+                      case HM.lookup fn idefs of
+                        Just _ -> return (Left fn)
+                        Nothing -> evalError (_nInfo f) ("Cannot resolve \"" ++ show f ++ "\"")
+                  (Nothing,_) -> evalError (_nInfo f) ("Cannot resolve \"" ++ show f ++ "\"")
+        return (d', dn, mapMaybe (either Just (const Nothing)) $ toList d')
+  sorted <- forM cs $ \c -> case c of
+              AcyclicSCC v -> return v
+              --CyclicSCC [v] -> return v
+              CyclicSCC vs -> do
+                evalError (if null vs then info else _tInfo $ view _1 $ head vs) $ "Recursion detected: " ++ show vs
+  let unifiedDefs :: Eval e (HM.HashMap Text Ref)
+      unifiedDefs = foldl dresolve (pure HM.empty) sorted
+      -- insert a fresh Ref into the map, fmapping the Either to a Ref via 'unify'
+      dresolve :: Eval e (HM.HashMap Text Ref) -> (Term (Either Text Ref), Text, [Text]) -> Eval e (HM.HashMap Text Ref)
+      dresolve eds (d,dn,_) = do
+        ds <- eds
+        ud <- traverse (unify info ds) d 
+        return $ HM.insert dn (Ref ud) ds
+      evalConstRef = runPure . evalConsts
+  defs <- unifiedDefs
+  evaluatedDefs <- traverse evalConstRef defs
+  let md = ModuleData i evaluatedDefs
+  installModule md
+  (evalRefs . rsNewModules) %= HM.insert _interfaceName md
+  return (gas1, idefs)
 
+    
 
+-- | Evaluate interface constraints in module 
+--
+-- For each implemented member, there may be overlapping meta
+-- models between the module implementation and the interface
+-- definition - therefore, in order not to step on the user's
+-- implementation, we must concatenate meta constraints and
+-- build the more detailed meta model of the member
+evaluateConstraints
+  :: Module
+  -> HM.HashMap Text Ref
+  -> Info
+  -> Eval e (HM.HashMap Text Ref)
+evaluateConstraints Interface{} _ info =
+  evalError info $ "Impossible. interface found while appending meta-constraints to module"
+evaluateConstraints Module{..} evalMap info = foldMap (evaluateConstraint evalMap info) _mInterfaces
+  where
+    evaluateConstraint :: HM.HashMap Text Ref -> Info -> ModuleName -> Eval e (HM.HashMap Text Ref)
+    evaluateConstraint hm i ifn = do
+      -- load the interface refmaps via refstore
+      iRefs <- preview $ eeRefStore . rsModules . ix ifn . mdRefMap
+      case iRefs of
+        -- if nothing found, interface is not loaded, ergo not unfound
+        Nothing -> evalError info $
+          "Interface implemented in module, but not defined: <" ++ asString' ifn ++ ">"
+        Just iRefs' -> HM.foldrWithKey (solveConstraint i) (pure hm) iRefs'
+
+-- | Compare implemented member signatures and concatenate models
+--
+-- For each reference, we must check that it exists in the interface refmap,
```

Comment:
Again, comments shouldn't say "how" but "why" or even "what". The first line here is sufficient, unless there's something you think isn't obvious in the below code.

Reply:
I've pared them down to the first line